aerospike/check.py

```diff
@@ -1,77 +1,205 @@
+# of the BSD-3 license. See the LICENSE file for details.
+
 
 # stdlib
 import socket
+import time
+import re
 from contextlib import closing
+from collections import namedtuple
 
 # project
 from checks import AgentCheck
 
-EVENT_TYPE = SOURCE_TYPE_NAME = 'aerospike'
-SERVICE_CHECK_NAME = '%s.server_up' % EVENT_TYPE
+SOURCE_TYPE_NAME = 'aerospike'
+SERVICE_CHECK_NAME = '%s.cluster_up' % SOURCE_TYPE_NAME
+CLUSTER_METRIC_TYPE = SOURCE_TYPE_NAME
+NAMESPACE_METRIC_TYPE = '%s.namespace' % SOURCE_TYPE_NAME
+NAMESPACE_TPS_METRIC_TYPE = '%s.namespace.tps' % SOURCE_TYPE_NAME
+SINDEX_METRIC_TYPE = '%s.sindex' % SOURCE_TYPE_NAME
+SET_METRIC_TYPE = '%s.set' % SOURCE_TYPE_NAME
+
+Addr = namedtuple('Addr', ['host', 'port'])
+
+def parse_namespace(data, namespace, secondary):
+    idxs = []
+    while data != []:
+        l = data.pop(0)
+
+        # $ asinfo -v 'sindex/phobos_sindex'
+        # ns=phobos_sindex:set=longevity:indexname=str_100_idx:num_bins=1:bins=str_100_bin:type=TEXT:sync_state=synced:state=RW; \
+        # ns=phobos_sindex:set=longevity:indexname=str_uniq_idx:num_bins=1:bins=str_uniq_bin:type=TEXT:sync_state=synced:state=RW; \
+        # ns=phobos_sindex:set=longevity:indexname=int_uniq_idx:num_bins=1:bins=int_uniq_bin:type=INT SIGNED:sync_state=synced:state=RW;
 
+        # $ asinfo -v 'sets/bar'
+        # ns=bar:set=demo:objects=1:tombstones=0:memory_data_bytes=34:truncate_lut=0:stop-writes-count=0:set-enable-xdr=use-default:disable-eviction=false;\
+        # ns=bar:set=demo2:objects=123456:tombstones=0:memory_data_bytes=8518464:truncate_lut=0:stop-writes-count=0:set-enable-xdr=use-default:disable-eviction=false
+
+        match = re.match('^ns=%s:([^:]+:)?%s=([^:]+):.*$' % (namespace, secondary), l)
+        if match == None:
+            continue
+        idxs.append(match.groups()[1])
+
+    return idxs   
 
 class AerospikeCheck(AgentCheck):
 
     def __init__(self, name, init_config, agentConfig, instances=None):
         AgentCheck.__init__(self, name, init_config, agentConfig, instances)
-        self.connection_pool = {}
+        self.connections = {}
 
     def check(self, instance):
-        """Run the Aerospike check for one instance."""
+        addr, metrics, namespace_metrics, required_namespaces, tags = \
+            self._get_config(instance)
 
-        # Required parameters
-        host = instance['host']
-        port = int(instance['port'])
+        try:
+            conn = self._get_connection(addr)
 
-        # Optional parameters
-        want_metrics = set(instance.get('metrics', []))
-        want_namespace_metrics = set(instance.get('namespace_metrics', []))
-        namespaces = instance.get('namespaces', None)
+            with closing(conn.makefile('r')) as fp:
+                conn.send('statistics\r')
+                self._process_data(fp, CLUSTER_METRIC_TYPE, metrics, tags=tags)
+
+                namespaces = self._get_namespaces(conn, fp, required_namespaces)
+
+                for ns in namespaces:
```

Comment:
How large can `namespaces` get? Trying to get an idea of the cardinality for the metric.... What could one expect `len(namespaces)` to be in a production cluster? Is it arbitrarily high? If so, we might have to set a soft cap.

Reply:
https://www.aerospike.com/docs/guide/limitations.html

Specifically, Limit of 32 Namespaces (databases), 1023 sets (tables), and 32767 bins (columns).  We don't usually see more than 4 in production clusters.

Reply:
perfect, nothing to worry about here.